fix bug #2123:Do not raise exception if socket is busy in TCPROSTransport

jinmenglei opened the pull request:
It's the same as roscpp socket.recv Exception handling process.

Commit:
fix bug #2123

It's the same as roscpp socket.recv Exception handling process.

fujitatomoya commented on clients/rospy/src/rospy/impl/tcpros_base.py:
```diff
@@ -100,12 +102,19 @@ def recv_buff(sock, b, buff_size):
+        else: #bomb out
+            raise TransportTerminated("unable to receive data from sender, check sender's logs for details")
+    except socket.error as ex:
+        if ex.args[0] in (EAGAIN, EWOULDBLOCK):
```
why it compares error message against error numbers? it is supposed to be `ex.errno`?

jinmenglei replied:
This place is reference C++
// never read more than INT_MAX since this is the maximum we can report back with the current return type
  uint32_t read_size = std::min(size, static_cast<uint32_t>(INT_MAX));
  int num_bytes = ::recv(sock_, reinterpret_cast<char*>(buffer), read_size, 0);
  if (num_bytes < 0)
  {
	if ( !last_socket_error_is_would_block() ) // !WSAWOULDBLOCK / !EAGAIN && !EWOULDBLOCK
    {
      ROSCPP_LOG_DEBUG("recv() on socket [%d] failed with error [%s]", sock_, last_socket_error_string());
      close();
    }
    else
    {
      num_bytes = 0;
    }
  }
  else if (num_bytes == 0)
  {
    ROSCPP_LOG_DEBUG("Socket [%d] received 0/%u bytes, closing", sock_, size);
    close();
    return -1;
  }

jinmenglei replied:
i change “ex.args[0]” to "ex.errno"

fujitatomoya commented on clients/rospy/src/rospy/impl/tcpros_base.py:
```diff
@@ -100,12 +102,19 @@ def recv_buff(sock, b, buff_size):
+            raise TransportTerminated("unable to receive data from sender, check sender's logs for details")
+    except socket.error as ex:
+        if ex.args[0] in (EAGAIN, EWOULDBLOCK):
+            raise TransportTerminated("unable to receive data from sender, check sender's logs for details")
```
if either EAGAIN or EWOULDBLOCK, it would want to retry. but raising exception with this particular message, it will break out the while loop in `receive_loop`.

jinmenglei replied:
I fixed this bug and had to deal with the logic

fujitatomoya commented on clients/rospy/src/rospy/impl/tcpros_base.py:
```diff
@@ -100,12 +102,19 @@ def recv_buff(sock, b, buff_size):
-        return len(d)
-    else: #bomb out
-        raise TransportTerminated("unable to receive data from sender, check sender's logs for details")
+    # add #2123
```
i think we can delete this comment.

jinmenglei replied:
i will delete it

fujitatomoya commented on clients/rospy/src/rospy/impl/tcpros_base.py:
```diff
@@ -806,6 +815,9 @@ def receive_loop(self, msgs_callback):
                     except:
                         pass
                     self.socket = None
+                    # add #2123
```
same here.

jinmenglei replied:
i delete this comment

fujitatomoya reviewed (commented):
IMO, I was thinking that where we would want to change is `receive_loop` to catch exception `TransportTerminated` before `TransportException`.

Commit:
fix bug #2123

delete comment

fix EAGAIN, EWOULDBLOCK logic

jinmenglei commented:
@fujitatomoya I didn't find a good way to deal with `transportterminated` before `TransportException`.Do you have a better solution?

fujitatomoya commented:
how about the following?

```
diff --git a/clients/rospy/src/rospy/impl/tcpros_base.py b/clients/rospy/src/rospy/impl/tcpros_base.py
index 57f340f49..b4bfd402b 100644
--- a/clients/rospy/src/rospy/impl/tcpros_base.py
+++ b/clients/rospy/src/rospy/impl/tcpros_base.py
@@ -800,6 +800,11 @@ class TCPROSTransport(Transport):
                     else:
                         self._reconnect()

+                except TransportTerminated as e:
+                    logerr("[%s] failed to receive incoming message : %s"%self.name, str(e))
+                    rospyerr("[%s] failed to receive incoming message: %s"%self.name, traceback.format_exc())
+                    break
+
                 except TransportException as e:
                     # set socket to None so we reconnect
                     try:
```

Commit:
fix bug #2123

add TransportTerminated before TransportException

delete if str(e)xxx

fix bug "TypeError: not enough arguments for format string" in DeserializationError

jinmenglei commented:
@fujitatomoya
I updated the code. Is this branch no longer maintained？
![image](https://user-images.githubusercontent.com/32763080/107170438-49c09e00-69fb-11eb-87f5-84ef0ec14569.png)

jinmenglei commented:
Is this PR to be closed because #2131 it has been merged?
@jacobperron

fujitatomoya commented:
http://wiki.ros.org/Distributions, kinetic is already E.O.L. I guess backport is not doable. so maybe just close this.

jinmenglei commented:
> http://wiki.ros.org/Distributions, kinetic is already E.O.L. I guess backport is not doable. so maybe just close this.

ok

jacobperron commented:
I will look into backporting #2131 to Melodic, following the next Noetic release.
